### PR TITLE
Multi Preview Fix 

### PR DESCRIPTION
**Pull request by daschuer**
This fixes an issue when you preview more than one cues at a time. 
In 2.3 the last released button moves the playposition to the associated hot cue position. 
This can't no longer work with cue loops involved. 

I have changed the code that internally only one hot cue can be previewing. 
If you preview one hot-cue and press a second one, the second takes over. The release of the first hot cue is ignored from that point. If it was a loop, looping is disabled.  

This also cleans up looking.

### CODE REVIEW

**Comment by uklotzde** [comment]
You might consider to pick and apply this commit: https://github.com/uklotzde/mixxx/commit/4c2d90f9e12acb49601db6819c964c41ba54e26d

I didn't check if it still applies here and what changes are required. The idea is simple, just keep the code that belongs together at a single place.

**Comment by daschuer** [comment]
Done.

File: src/engine/controls/cuecontrol.cpp [hunk 1]
```diff
@@ -30,6 +30,8 @@ constexpr double CUE_MODE_CUP = 5.0;
 /// This is the position of a fresh loaded tack without any seek
 constexpr double kDefaultLoadPosition = 0.0;
 constexpr int kNoHotCueNumber = 0;
+/// Used for a common tarcking of the previewing Hotcue in m_currentlyPreviewingIndex
```
Comment (uklotzde) [inline src/engine/controls/cuecontrol.cpp:33]: typo

File: src/engine/controls/cuecontrol.cpp [hunk 2]
```diff
@@ -416,10 +416,16 @@ void CueControl::detachCue(HotcueControl* pControl) {
     pControl->resetCue();
 }
 
+// This is called from the EngineWokerThread and ends with the initial seek v
```
Comment (uklotzde) [inline src/engine/controls/cuecontrol.cpp:419]: typo and left over character

File: src/engine/controls/cuecontrol.cpp [hunk 3]
```diff
@@ -617,6 +596,14 @@ void CueControl::loadCuesFromTrack() {
         }
     }
 
+    // Detach all hotcues that are no longer present
+    for (int hotCue = 0; hotCue < m_iNumHotCues; ++hotCue) {
+        if (!active_hotcues.contains(hotCue)) {
```
Comment (uklotzde) [inline src/engine/controls/cuecontrol.cpp:601]: Fix the variable naming along the way?

File: src/engine/controls/cuecontrol.cpp [hunk 4]
```diff
@@ -1269,10 +1175,12 @@ void CueControl::cueGoto(double value) {
         return;
     }
 
-    QMutexLocker lock(&m_mutex);
+    QMutexLocker lock(&m_trackMutex);
     // Seek to cue point
     double cuePoint = m_pCuePoint->get();
 
+    // Note: We do not mess with ply here, we continue playing or previewing.
```
Comment (uklotzde) [inline src/engine/controls/cuecontrol.cpp:1182]: typo

File: src/engine/controls/cuecontrol.cpp [hunk 5]
```diff
@@ -1640,14 +1519,20 @@ void CueControl::introEndSet(double value) {
     TrackPointer pLoadedTrack = m_pLoadedTrack;
     lock.unlock();
 
+    // Update Track's cue.
+    // CO's are update loadCuesFromTrack()
```
Comment (uklotzde) [inline src/engine/controls/cuecontrol.cpp:1523]: missing "in" (multiple times)

File: src/engine/controls/cuecontrol.cpp [hunk 6]
```diff
@@ -2218,16 +2122,15 @@ void CueControl::slotLoopUpdated(double startPosition, double endPosition) {
...
-    }
-
     if (m_pCurrentSavedLoopControl->getStatus() != HotcueControl::Status::Active) {
         slotLoopReset();
         return;
     }
 
-    CuePointer pCue(m_pCurrentSavedLoopControl->getCue());
+    CuePointer pCue = m_pCurrentSavedLoopControl->getCue();
+    if (!pCue) {
```
Comment (uklotzde) [inline src/engine/controls/cuecontrol.cpp:2131]: This was never checked before. Could this ever become a nullptr? If not then use VERIFY_OR_DEBUG_ASSERT.

File: src/engine/controls/cuecontrol.h [hunk 7]
```diff
@@ -98,18 +99,21 @@ class HotcueControl : public QObject {
...
-        m_previewingType = type;
+        return m_previewingType.getValue();
     }
     double getPreviewingPosition() const {
-        return m_previewingPosition;
+        return m_previewingPosition.getValue();
     }
-    void setPreviewingPosition(double position) {
-        m_previewingPosition = position;
+    void storePreviewingActivateData() {
```
Comment (uklotzde) [inline src/engine/controls/cuecontrol.h:110]: The name of the function does not tell me what it is actually doing. What is "activate data"?

File: src/engine/controls/cuecontrol.h [hunk 8]
```diff
@@ -335,24 +339,14 @@ class CueControl : public EngineControl {
...
-    // We have multiple cases where it locks m_pLoadedTrack and
-    // pControl->getCue(). This guards the hotcueClear() that could detach the
-    // cue call, but doesn't protect from cue changes via loadCuesFromTrack()
-    // which is called outside the mutex lock.
-    //
-    // We need to repair this.
-    QMutex m_mutex;
+    TrackPointer m_pLoadedTrack; // is written from an engine worker thread
+    // Must be locked when using the m_pLoadedTrack and it's properties
+    QMutex m_trackMutex;
```
Comment (uklotzde) [inline src/engine/controls/cuecontrol.h:349]: Place the mutex above the member(s) it protects as a delimiter after the unprotected members.

File: src/library/rekordbox/rekordboxfeature.cpp [hunk 9]
```diff
@@ -794,29 +794,29 @@ void setHotCue(TrackPointer track,
...
-        pCue = CuePointer(track->createAndAddCue());
+    mixxx::CueType type = mixxx::CueType::HotCue;
+    if (endPosition != Cue::kNoPosition) {
+        type = mixxx::CueType::Loop;
     }
 
-    pCue->setStartPosition(startPosition);
-    if (endPosition == Cue::kNoPosition) {
-        pCue->setType(mixxx::CueType::HotCue);
+    if (!pCue) {
```
Comment (uklotzde) [inline src/library/rekordbox/rekordboxfeature.cpp:810]: if (pCue) { update } else { create }

File: src/mixer/basetrackplayer.cpp [hunk 10]
```diff
@@ -327,11 +327,14 @@ TrackPointer BaseTrackPlayerImpl::unloadTrack() {
             }
         }
         if (!pLoopCue) {
```
Comment (uklotzde) [inline src/mixer/basetrackplayer.cpp:329]: if (pLoopCue) { update } else { create }

File: src/track/track.cpp [hunk 11]
```diff
@@ -813,9 +815,17 @@ void Track::slotCueUpdated() {
     emit cuesUpdated();
 }
 
-CuePointer Track::createAndAddCue() {
+CuePointer Track::createAndAddCue(
+        mixxx::CueType type,
+        int hotCueIndex,
+        double sampleStartPosition,
+        double sampleEndPosition) {
     QMutexLocker lock(&m_qMutex);
```
Comment (uklotzde) [inline src/track/track.cpp:823]: Move the lock down to the point when it is actually needed. We don't need to protect the global system allocator or the Qt connect functions.

**Review by uklotzde — CHANGES_REQUESTED** [review]
Just some minor coding comments. Overall the code seems to be much more readable than before.

I didn't check if everything works as expected. The cue handling is complicated, I have to trust the tests.

**Comment by Be-ing** [comment]
To clarify, this fixes a regression in the main branch, correct? If so, it's not relevant for 2.3.

**Comment by daschuer** [comment]
Yes, merge target main is correct.

**Comment by daschuer** [comment]
Done

**Comment by daschuer** [comment]
@uklotzde Is this OK now?

**Review by uklotzde — APPROVED** [review]
Sorry for the delay, didn't notice. LGTM

**Pull request merged**